File: java/zijietiaodong/str/LongestCommonPrefix.java
package str;

public class LongestCommonPrefix {

    public String longestCommonPrefix(String[] strs) {
        if (strs.length == 0) return "";
        return divideAndConquer(strs, 0, strs.length - 1);
    }

    public String divideAndConquer(String[] strs, int start, int end) {
        if (start == end) {
            return strs[start];
        }
        int mid = (start + end) / 2;
        String left = divideAndConquer(strs, start, mid);
        String right = divideAndConquer(strs, mid + 1, end);
        return getCommonPrefix(left, right);
    }

    public String getCommonPrefix(String a, String b) {
        int i = 0, j = 0;
        while (i < a.length() && j < b.length()) {
            if (a.charAt(i) == b.charAt(j)) {
                i++;
                j++;
            } else {
                break;
            }
        }
        return a.substring(0, i);
    }

    public static void main(String[] args) {
        System.out.println(new LongestCommonPrefix().longestCommonPrefix(new String[]{ "flower","flow","flight" }));
    }
}

File: go/tencent/array_and_str/longest_common_prefix.go
package main

import "fmt"

func main() {
	fmt.Println(longestCommonPrefix([]string{"flower", "flow", "flight"}))
}

func longestCommonPrefix(strs []string) string {
	return merge(strs, 0, len(strs)-1)
}

func merge(strs []string, i int, j int) string {
	if i > j {
		return ""
	}
	if i == j {
		return strs[i]
	}
	mid := (i + j) / 2
	left, right := merge(strs, i, mid), merge(strs, mid+1, j)
	return getCommonPrefixOfTwoStr(left, right)
}

func getCommonPrefixOfTwoStr(s1 string, s2 string) string {
	i := 0
	for ; i < len(s1) && i < len(s2); i++ {
		if s1[i] != s2[i] {
			break
		}
	}
	return s1[:i]
}
